### ISSUE TITLE
Detected framerate is wrong

### ISSUE THREAD

**Issue by OParczyk** [WyattBlue/auto-editor#140]
**What went wrong**
I tried to auto-edit a video and noticed a steadily increasing desync in the output. No unusual terminal output here. Further investigation yielded a wrongly detected framerate.

**Does the example video work? (if applicable)**
Doesn't happen with the example video.

**What does the program output when `--debug` is appended to your command/use case**

```
   - Temp Directory: /tmp/tmptimxl1_3
   - Speeds: [99999, 1.0]                                                                                                                                                                                                                    
   - INPUT_FILE: test.mkv
   - newOutput: test_ALTERED.mkv
   - sampleRate: 48000
   - audioBitrate: None
Frame rate: 30.303030303030305
   - Frame Rate: 30.303030303030305
Analyzing audio volume.
  ⏳Creating new audio: [██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████] 100% done ETA 06:10 PM                
   - Total Samples: 96321223
   - Samples per Frame: 1584.0
   - Expected video length: 60808.8529040404
Using av method
   - pix_fmt: yuv420po: [░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░] 0% done ETA 06:10 PM                  
Writing the output file.[████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████░░] 99.0% done ETA 06:18 PM               
Finished. took 529.12 seconds (0:08:49)                                                                                                                                                                                                      
Warning! Could not open output file.
Deleting temp dir

```

**What's your platform? (if applicable)**

```
Python Version: 3.6.12 64-bit
Platform: Linux 5.4.0-65-generic
Config File path: /home/user/.local/lib/python3.6/site-packages/auto_editor/config.txt
FFmpeg path: ffmpeg
FFmpeg version: 4.2.4-1ubuntu0.1
Auto-Editor version 21w05c
```

**What's your input? (if applicable)**
Actual framerate is 30.0 fps, for more detail on the input see this: [mediainfo.log](https://github.com/WyattBlue/auto-editor/files/5973314/mediainfo.log).
Output of `auto-editor info test.mkv`:
```
file: test.mkv
 - fps: 30.303030303030305
 - duration: N/A
 - resolution: 1920x1080
 - video codec: hevc
 - video bitrate: N/A
 - audio tracks: 1
   - Track #0
     - codec: flac
     - samplerate: 48.0 kHz
     - bitrate: N/A
```

My current workaround is to first convert the video to h.264, then auto-edit and convert the result again. However quality degradation is noticable and this process takes a lot of time.

**Comment by OParczyk** [comment]
This effect may also be reproduced with the video acquired using
`youtube-dl -f bestvideo[ext=mp4]+bestaudio[ext=m4a] https://www.youtube.com/watch?v=jNQXAC9IVRw`
and running `python3 -m auto_editor "Me at the zoo-jNQXAC9IVRw.mp4" --debug`.
a framerate of 15.053003533568905 fps is reported, while the actual framerate is 15.000 fps.

A weird thing of this file is that the duration of audio and video channels seems to mismatch, see [mediainfo.txt](https://github.com/WyattBlue/auto-editor/files/5984360/mediainfo.txt) for details. This isn't the case in the original Issue file however. Also, both audio and video codecs are different.

I'm using version 2021.01.24.1 of `youtube-dl` in case that's relevant.

**Comment by OParczyk** [comment]
I found another extreme case, also acquired using `youtube-dl -f bestvideo[ext=mp4]+bestaudio[ext=m4a]`. I'm uncertain, whether or not I am allowed to share the video URL publicly, but I think I can share the URL with you privately, if you tell me a means by which you'd like to receive it.
`auto-edit info test.mp4` yielded this:
```
file: test.mp4
 - fps: 3.0
 - duration: 1285.033333
 - resolution: 1280x720 (16:9)
 - video codec: h264
 - video bitrate: 122k
 - audio tracks: 1
   - Track #0
     - codec: aac
     - samplerate: 44.1 kHz
     - bitrate: 128k
```
The detected fps is _way_ off, in reality more like 30. For details see this: [mediainfo.log](https://github.com/WyattBlue/auto-editor/files/5984423/mediainfo.log)

`auto-edit` yields a very laggy and jumpy video, where the video and audio are completely desynced.
you'll need 09efd954f5c0de22ecb7a92efb26d81294473497 for this and the above comment, because both videos are affected by #138.

**Comment by WyattBlue** [comment]
@OParczyk Send over the video to wyattbluesandbox@gmail.com if you can, otherwise, run the command `ffprobe -show_streams out.mp4` on your video and show the output. 

Try seeing if adding `metadataOps --force_fps_to 30` will stop the desync. That will override auto-editors guess at the fps with that value.

If you didn't already guess, fps is a very important value when trying to line up video and audio, so an issue with detecting the fps would lead to desync issues.

**Comment by OParczyk** [comment]
I've sent you an Email.
This is the output of the given `ffprobe` command: [ffprobe.txt](https://github.com/WyattBlue/auto-editor/files/6012528/ffprobe.txt)
`metadataOps --force_fps_to 30` works for https://github.com/WyattBlue/auto-editor/issues/140#issuecomment-779480597 and the original video in the issue. ([`ffprobe` for that](https://github.com/WyattBlue/auto-editor/files/6012641/ffprobe2.txt))

Let's ignore https://github.com/WyattBlue/auto-editor/issues/140#issuecomment-779473717 going forward as per https://github.com/WyattBlue/auto-editor/issues/142#issuecomment-779526563.

**Comment by OParczyk** [comment]
I've digged into the latter `ffprobe` output a little and I'm surprised how the least common multiple of all present framerates `r_frame_rate` can be less than the average framerate `avg_frame_rate`. This being a constant framerate video file makes that only stranger.
As per [this stackexchange thread](https://video.stackexchange.com/questions/20789/ffmpeg-default-output-frame-rate/20790) `ffmpeg` uses `r_frame_rate` when creating constant framerate output and IMHO doing so (and doubling frames if necessary) is quite a good idea. The issue here suggests that auto-editor uses `avg_frame_rate` so maybe going with `r_frame_rate` would be worth a thought.

**Comment by OParczyk** [comment]
fcba1f48a0d1df8afd81ed251ad26307e1e189dc Fixed the original issue here. https://github.com/WyattBlue/auto-editor/issues/140#issuecomment-779480597 however is still way off. Although both `r_frame_rate` and `avg_frame_rate` are `30/1`, auto-editor recognizes ~0.0002 fps and crashes soon after:

```
$ python3.9 -m auto_editor test.mp4 --debug

   - Temp Directory: /tmp/tmpcfayoyl5
   - Speeds: [99999, 1.0]                                                                                                                                                                                                                    
   - INPUT_FILE: test.mp4
   - newOutput: test_ALTERED.mp4
   - sampleRate: 44100
   - audioBitrate: 128k
Frame rate: 0.00015789473684210527
   - Frame Rate: 0.00015789473684210527
Analyzing audio volume.
Traceback (most recent call last):
  File "/usr/lib/python3.9/runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.9/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/home/user/from_source/auto-editor/auto_editor/__main__.py", line 692, in <module>
    main()
  File "/home/user/from_source/auto-editor/auto_editor/__main__.py", line 579, in main
    chunks = applySpacingRules(hasLoud, fps, args.frame_margin,
  File "/home/user/from_source/auto-editor/auto_editor/cutting.py", line 290, in applySpacingRules
    return chunkify(includeFrame, arrayLen)
  File "/home/user/from_source/auto-editor/auto_editor/cutting.py", line 253, in chunkify
    chunks.append([startP, arrayLen, int(includeFrame[j])])
UnboundLocalError: local variable 'j' referenced before assignment
```

**Comment by OParczyk** [comment]
a9a5897811c103d3d6811646f43bf3d3ebddb03a fixes above crash, but not the issue. `ffprobe` reports the framerate to be `30/1`, so why does auto-edit think its below 1?

**Comment by WyattBlue** [comment]
I don't know why auto-editor thinks the fps is drastically different. I experimented on the video that you sent to my email and "test4.mkv", and the input and output was always 30 fps. 

I did notice that the output for "test4.mkv" that the in the last part, it the video's timer on "20.25". I changed the input's framerate to 30 via ffmpeg but I got the same result. I then changed the framerate to 15 and it didn't get stuck. I also "fixed" it by changing the container from mkv to mp4.

I looked into it more and realized that test4.mkv has a Variable Frame Rate.

```
ffmpeg -i /Users/wyattblue/Downloads/test4.mkv   -vf vfrdet -an -f null -
```

```
[Parsed_vfrdet_0 @ 0x7fc91c50cf80] VFR:0.679155 (707/334) min: 33 max: 133 avg: 40
```

I believe that means that around 68% of the video is varied, and the constant fps portion is about 20fps and these average to about 30fps. 

Auto-Editor can't handle variable frame rates, which is why there's desync when it tries to edit VFR files. 
I'm 99% sure variable frame rates are the  **only** reason why it's desyncing and causing issues, but these files still need to be detected and hopefully, be reprocessed to a constant frame rate. That's why I'll continue this conversation to your desync issue.

`auto-editor info` should display whether or not a video has VFR, so I'm implement that in a few days.
